### ISSUE TITLE
Support for External Frameworks like Tensorflow.js

### ISSUE THREAD

**Issue by krumiaa** [ncsoft/Unreal.js#226]
Ran into errors getting this to work with Tensorflow.js, specifically SyntaxError: Unexpected token * when using import * as tf from '@tensorflow/tfjs'; to load the tensor flow. Is Unreal JS fundamentally incompatible with external packages like Tensorflow or am I missing something?

**Comment by crocuis** [comment]
Use an api exposed to Unreal such as https://github.com/getnamo/tensorflow-ue4.

**Comment by keicoon** [comment]
@krumiaa Hi. Do you find any solution? I writed some code that you want to use 'TFJS'. Check my repo(https://github.com/keicoon/Deep-Learning/tree/master/DQN). It has dependencies about unreal.js and @tensorflow/tfjs.

**Comment by krumiaa** [comment]
Did you get this DQN repo to work with Unreal.js? When I tried to load tfjs as a dependency, it threw a bunch of errors....

**Comment by krumiaa** [comment]
Still not sure how you got this working, was it with a node install of tensorflow.js or via npm? Trying to run example-dqn-tfjs.js from your DQN repository, kept pulling errors like module ReferenceError: module is not defined or ReferenceError:process is not defined. Is it possible you could share the unreal project file or do a step by step run through of the install. I tried it on Unreal 4.19

**Comment by keicoon** [comment]
@krumiaa Okay. Let me check again:)

**Comment by keicoon** [comment]
@krumiaa Okay. Let me check again :)

**Comment by krumiaa** [comment]
Tried the project from getlab as well as the new github code but getting error: Cannot read property 'train' of undefined. See image. Thinking this maybe some version specific glitch. Am using npm 6.4.1 and Unreal.js ver 0.3.1 with Unreal 4.19. I didnt see that tensorflow was installed in the scripts directory of your project so i tried npm install @tensorflow/tfjs to the /scripts but still not working. Any other 
suggestions for how to replicate this? Maybe I'm missing something obvious?
![dqn_error](https://user-images.githubusercontent.com/5367529/47510140-26c74b80-d83d-11e8-8a02-45501ff0492a.jpg)

**Comment by keicoon** [comment]
Hi!
I am using tensorflow.js bundle file that made by `browserify`. It will be generated by `Scripts/bundle_tensorflow_tfjs.js`. So, you first check `bundle_tensorflow_tfjs.js` file be existed. If not existed, Run below command repeatedly until `tf` is not empty.
```
tf = require('dqn\lib\fancyLoadNPM')('@tensorflow/tfjs', global)
console.log(tf)
```
If you don't want to follow, you can also make bundle file using `browserify` by independent command.
Thank YOU :)

**Comment by krumiaa** [comment]
Thank you so much for helping explain this. After hours and hours, I still have not managed to get it working. Do I need browserfy installed in order to run the project as you have made it? the only  way i get console.log(tf) to rerturn anything is with const tf = require('/dqn/lib/fancyLoadNPM')('bundle_tensorflow_tfjs.js', global); but this creates multiple versions of bundle_tensorflow_tfjs.js in the scripts directory and basic tensorflow functions like const model = tf.sequential(); still wont work. Any change u could post some stimple step by step setup instructions for loading all the dependencies on the github page?

**Comment by keicoon** [comment]
Yes, this solution need to install nodejs that he said. Because, I use command(`JavascriptProcess.Create('node', ...)`) that has dependency about nodejs. Let's install [nodejs](https://nodejs.org/en/) 8.x.x. :)

**Comment by krumiaa** [comment]
Were you able to get this working Jan? Iv been banging my head against it for a week. Yes, have node 8.12 installed. Wondering if there are some build path issues with the fancyLoadNPM code, parts that have to be modified depending on where node was installed or where game assets are located. Or perhaps dependent on using a particular version of node. Cant think of any other reasons the project file wouldn’t work out of the box. Or perhaps Browserify must be installed at a particular location, noticed its referenced in fancyLoadNPM. Allot of moving parts. Not even sure why  FancyLoadNPM is required, what purpose it is serving here different from NPM. Would be helpful if some of this was laid out in the documentation, though I suppose that would take away all the fun ;-)

**Comment by keicoon** [comment]
Simply, I would explain why need to use `browerify` and `node` that not `v8`. First unreal.js not support npm module officially so we can't use `@tensorflow.js` directly. And v8 is different to node about loading module system has circularly issue. So, I use trick will resolve circular dependency module issue using `browserify`. Lastly, I will upload bundle script that be browserified about tensorflow.js for you :)

**Comment by krumiaa** [comment]
Very much appreciate the work you have done on this keicoon and believes it really opens up some cool possibilities for doing machine learning in Unreal.  Really confused on how to replicate though so anything you can do to make the process simpler, such as the bundle script you mentioned, would be awesome.

**Comment by keicoon** [comment]
@krumiaa Hi!
Download this file[bundle_tensorflow_tfjs.js](https://github.com/keicoon/Deep-Learning/releases/download/0.1.0/bundle_tensorflow_tfjs.js).
and put it in `Content\Scripts`. :)

**Comment by krumiaa** [comment]
Hi keicoon, thanks for uploading this. I noticed a similar bundle file in your unreal project scripts directory, wasn't sure what the the proper way to deploy it is. Does one still need the fancyLoadNPM to access this bundle file or can I just call it like const tf = require('bundle_tensorflow_tfjs', global);  Thanks for clarifying

**Comment by Master244** [comment]
I got this to work as well very cool! Didn't even think about possibilities with tensorflow etc cool stuff! I hope to share some cool stuff soon.
